cdap-kafka/src/main/java/co/cask/cdap/kafka/run/KafkaServerMain.java

```diff
@@ -102,11 +94,25 @@ public void init(String[] args) {
       }
     }
 
-    int brokerId = generateBrokerId(address);
-    LOG.info(String.format("Initializing server with broker id %d", brokerId));
+    kafkaProperties = generateKafkaConfig(cConf);
+
+    Preconditions.checkState(Integer.parseInt(
+      kafkaProperties.getProperty(KafkaConstants.ConfigKeys.NUM_PARTITIONS_CONFIG)) > 0,
+                             "Num partitions should be greater than zero.");
+    int port = Integer.getInteger(kafkaProperties.getProperty(KafkaConstants.ConfigKeys.PORT_CONFIG), -1);
+    Preconditions.checkState(port > 0, "Port number is invalid.");
 
-    kafkaProperties = generateKafkaConfig(brokerId, zkConnectStr, hostname, port, numPartitions,
-                                          replicationFactor, logDir);
+    if (kafkaProperties.get("broker.id") == null) {
+      int brokerId = generateBrokerId(address);
+      LOG.info(String.format("Initializing server with broker id %d", brokerId));
+      kafkaProperties.setProperty("broker.id", Integer.toString(brokerId));
+    }
+    if (hostname == null) {
+      kafkaProperties.setProperty("host.name", null);
```

Comment:
True. But when `HOSTNAME_CONFIG` holds 0.0.0.0, does the code below set the var `hostname` to null? I think that is the current issue.

```
    InetAddress address = Networks.resolve(hostname, new InetSocketAddress("localhost", 0).getAddress());
    if (address.isAnyLocalAddress()) {
      try {
        address = InetAddress.getLocalHost();
      } catch (UnknownHostException e) {
        throw Throwables.propagate(e);
      }
    }
    if (address.isLoopbackAddress()) {
      LOG.warn("Binding to loopback address!");
    }
    hostname = address.getCanonicalHostName();
```